packages/firestore/src/core/event_manager.ts

```diff
@@ -56,11 +60,21 @@ export class EventManager implements SyncEngineListener {
 
   private snapshotsInSyncListeners: Set<Observer<void>> = new Set();
 
-  constructor(private syncEngine: SyncEngine) {
-    this.syncEngine.subscribe(this);
+  /** Callback invoked when a Query is first listen to. */
+  private onListen?: (query: Query) => Promise<ViewSnapshot>;
+  /** Callback invoked once all listeners to a Query are removed. */
+  private onUnlisten?: (query: Query) => Promise<void>;
+
+  subscribe(
```

Comment:
Update: I had to revert this since this breaks the "terminate()" tests. The call to get the component provider throws an exception if the client has already been terminated, and converting these calls to async functions turns that into a rejected Promise.